hotshot-types/src/traits/election.rs

```diff
@@ -153,3 +157,78 @@ pub trait Membership<TYPES: NodeType>: Debug + Send + Sync {
         None
     }
 }
+
+/// Calculate `xor(drb.cycle(), public_key)`, returning the result as a vector of bytes
+fn cyclic_xor(drb: [u8; 32], public_key: Vec<u8>) -> Vec<u8> {
+    let drb: Vec<u8> = drb.to_vec();
+
+    let mut result: Vec<u8> = vec![];
+
+    for (drb_byte, public_key_byte) in public_key.iter().zip(drb.iter().cycle()) {
+        result.push(drb_byte ^ public_key_byte);
+    }
+
+    result
+}
+
+/// Generate the stake table CDF, as well as a hash of the resulting stake table
+pub fn generate_stake_cdf<Key: SignatureKey, Entry: StakeTableEntryType<Key>>(
+    mut stake_table: Vec<Entry>,
+    drb: [u8; 32],
+) -> (Vec<(Entry, U256)>, [u8; 32]) {
+    // sort by xor(public_key, drb_result)
+    stake_table.sort_by(|a, b| {
+        cyclic_xor(drb, a.public_key().to_bytes()).cmp(&cyclic_xor(drb, b.public_key().to_bytes()))
+    });
+
+    let mut hasher = Sha256::new();
+
+    let mut cumulative_stake = U256::from(0);
+    let mut cdf = vec![];
+
+    for entry in stake_table {
+        cumulative_stake += entry.stake();
+        hasher.update(entry.public_key().to_bytes());
+
+        cdf.push((entry, cumulative_stake));
+    }
+
+    (cdf, hasher.finalize().into())
+}
+
+/// select the leader for a view
+///
+/// # Panics
+/// Panics if the cdf is empty
+pub fn select_randomized_leader<SignatureKey, Entry: StakeTableEntryType<SignatureKey> + Clone>(
+    cdf: Vec<(Entry, U256)>,
+    stake_table_hash: [u8; 32],
+    drb: [u8; 32],
+    view: u64,
+) -> Entry {
+    // We hash the concatenated drb, view and stake table hash.
+    let mut hasher = Sha512::new();
+    hasher.update(drb);
+    hasher.update(view.to_le_bytes());
+    hasher.update(stake_table_hash);
+    let raw_breakpoint: [u8; 64] = hasher.finalize().into();
+
+    // then calculate the remainder modulo the total stake as a U512
+    let remainder: U512 =
+        U512::from_little_endian(&raw_breakpoint) % U512::from(cdf.last().unwrap().1);
+
+    // and drop the top 32 bytes, downcasting to a U256
+    let breakpoint: U256 = {
```

Comment:
Should we use this instead? 
https://docs.rs/primitive-types/latest/primitive_types/struct.U256.html#impl-TryFrom%3CU512%3E-for-U256